python/ray/data/impl/split.py

```diff
@@ -0,0 +1,1106 @@
+        # We give a custom node resource in order to ensure that this
+        # splitting task is scheduled on the same node as the block.
+        resource_label = get_resource_label(node_id)
+        did_split_for_node, offset = _split_block_over_actor_bins(
+            block,
+            block_set,
+            actor_bin_subset,
+            split_resources={resource_label: 0.001},
+            offset=offset)
+        did_split |= did_split_for_node
+    if did_split:
+        block_set.discard_block(block)
+        if offset < block.num_rows:
+            # There's a leftover slice of the block that wasn't allocated
+            # in this round, save it for allocation in the next round.
+            (leftover_block,
+             leftover_metadata) = _cached_split_block_at_indices().options(
+                 num_returns=2, resources={
+                     resource_label: 0.001
+                 }).remote(block.block, block.meta, [offset, block.num_rows])
+            leftover_size = block.num_rows - offset
+            leftover_block_data = BlockData(leftover_block, leftover_metadata,
+                                            leftover_size)
+            block_set.add_block(leftover_block_data)
+
+
+def _split_block_over_actor_bins(
+        block: BlockData,
+        block_set: BlockSet,
+        actor_bin_set: BinSet,
+        split_resources: Optional[Dict[str, float]] = None,
+        offset: Optional[int] = 0):
+    """Split the given block over the provided actor bin set, returning
+    the end-boundary offset if the entire block's contents aren't allocated.
+    """
+    split_indices, actor_bins, offset = _get_block_split_indices(
+        actor_bin_set, block.num_rows, offset)
+    if split_indices is None:
+        return False, offset
+    split_sizes = _get_split_sizes(split_indices)
+    # Submit splitting task.
+    block_splits = _create_splits(
+        block.block,
+        block.meta,
+        split_indices,
+        split_resources=split_resources)
+    for (split_block, split_meta), actor_bin, num_rows in zip(
+            block_splits, actor_bins, split_sizes):
+        block_data = BlockData(split_block, split_meta, num_rows)
+        block_set.add_block(block_data)
+        actor_bin.allocate(block_data)
+        block_set.mark_allocated(block_data)
+    return True, offset
+
+
+def split_blocks_over_arbitrary_actors(block_set: BlockSet,
+                                       actor_bin_set: BinSet):
+    """Split the given block set over the given actor set, with not subsetting
+    (locality) constraints.
+    """
+    # Get blocks sorted by size (increasing).
+    blocks = block_set.get_unallocated_blocks_ordered_by_size()
+    for block in blocks:
+        did_split, _ = _split_block_over_actor_bins(block, block_set,
+                                                    actor_bin_set)
+        assert did_split
+
+
+def rebalance_blocks(block_set: BlockSet, actor_bin_set: BinSet,
```

Comment:
is this necessary @ericl ?
Even if we need this, how about we just split the larger blocks? This should reduce the 200 lines of code to 10-20 lines of code.